Data obtained from `https://data.boston.gov/dataset/buildbps-facilities-and-educational-data-for-boston-public-schools`

In [1]:
import pandas as pd
import numpy as np

In [2]:
bos = pd.read_csv('../datasets/buildbps.csv')
bos

SMMA_Identifier SMMA_Only_For_Map                       BPS_School_Name  \
0               031               NaN              Adams, Samuel Elementary   
1               078               NaN    Alighieri, Dante Montessori School   
2              045A               NaN            Another Course to College*   
3               012               NaN  Baldwin Early Learning Pilot Academy   
4               087               NaN             Bates, Phineas Elementary   
..              ...               ...                                   ...   
136            011B               NaN                 West Roxbury Academy*   
137            047B               NaN      West Zone Early Learning Center*   
138             100               NaN          Winship, F. Lyman Elementary   
139             073               NaN             Winthrop, John Elementary   
140             044               NaN    Young Achievers Science & Math K-8   

            BPS_Historical_Name SMMA_Abbreviated_Name  \
0                         Adams                 Adams   
1                     Alighieri             Alighieri   
2                          Taft                  ACC*   
3                   Baldwin ELC               Baldwin   
4                         Bates                 Bates   
..                          ...                   ...   
136    West Roxbury High School         West Roxbury*   
137  Hennigan Elementary School            West Zone*   
138                     Winship               Winship   
139         Winthrop Elementary              Winthrop   
140                   Lewenberg       Young Achievers   

                               BPS_Address BRA_Neighborhood  SMMA_latitude  \
0     165 Webster St East Boston, MA 02128      East Boston      42.365553   
1     37 Gove Street East Boston, MA 02128      East Boston      42.371565   
2      20 Warren Street Brighton, MA 02135          Allston      42.350354   
3          121 Corey Rd Brighton, MA 02135         Brighton      42.342037   
4        426 Beech St Roslindale, MA 02131       Roslindale      42.277663   
..                                     ...              ...            ...   
136  1205 VFW Pkwy, West Roxbury, MA 02132     West Roxbury      42.282206   
137   200 Heath St Jamaica Plain, MA 02130    Jamaica Plain      42.326026   
138       54 Dighton St Brighton, MA 02135         Brighton      42.347723   
139   35 Brookford St Dorchester, MA 02125          Roxbury      42.318387   
140        20 Outlook Rd Mattapan, MA02126         Mattapan      42.281976   

     SMMA_longitude      SMMA_Typology  ... SMMA_EA_K8_Adequacy_Cafeteria  \
0        -71.034917  Elementary School  ...                           NaN   
1        -71.037608  Elementary School  ...                           NaN   
2        -71.145582        High School  ...                           NaN   
3        -71.140529     Early Learning  ...                           NaN   
4        -71.135353  Elementary School  ...                           NaN   
..              ...                ...  ...                           ...   
136      -71.174549        High School  ...                           NaN   
137      -71.106821     Early Learning  ...                           NaN   
138      -71.154965  Elementary School  ...                           NaN   
139      -71.075341  Elementary School  ...                           NaN   
140      -71.093909                K-8  ...                          Poor   

    SMMA_EA_K8_Adequacy_Stage SMMA_EA_K8_Adequacy_Medical  \
0                         NaN                         NaN   
1                         NaN                         NaN   
2                         NaN                         NaN   
3                         NaN                         NaN   
4                         NaN                         NaN   
..                        ...                         ...   
136                       NaN                         NaN   
137                 

### Columns

One of the greatest challenges in this notebook is going to be examining the columns, filling `NaaNs`, and choosing which columns should be dropped from this dataset. After all, there are **251** columns worth of information for just 141 schools.

One cannot view all the columns in this dataset simply by running the command `bos.columns` -- there are simply too many to display. Let's look at all the columns in batches of 50.

In [3]:
bos.columns[:50]

Index(['SMMA_Identifier', 'SMMA_Only_For_Map', 'BPS_School_Name',
       'BPS_Historical_Name', 'SMMA_Abbreviated_Name', 'BPS_Address',
       'BRA_Neighborhood', 'SMMA_latitude', 'SMMA_longitude', 'SMMA_Typology',
       'SMMA_Educational_Program', 'BPS_Grades_Offered', 'BPS_Property_Status',
       'BPS_Total_GSF', 'BPS_GSF', 'SMMA_Site_SF', 'SMMA_Site_Acres',
       'BPS_Year_Founded ', 'BPS_Year_Built', 'MSBA_Year_Reno', 'SMMA_Era',
       'BPS_Plans', 'BPS_Principal', 'BPS_Open', 'BPS_Close', 'DOE_ID',
       'DOE_Total', 'DOE_PK', 'DOE_K', 'DOE_1', 'DOE_2', 'DOE_3', 'DOE_4',
       'DOE_5', 'DOE_6', 'DOE_7', 'DOE_8', 'DOE_9', 'DOE_10', 'DOE_11',
       'DOE_12', 'DOE_SP', 'GSF / Student Value', 'BPS_KWH',
       ' BPS_Electric_Bill ', 'BPS_Therms', ' BPS_Gas_Bill ',
       'BPS_Water_ft_cubic', ' BPS_Water_Bill ', ' BPS_Total_Energy_Cost '],
      dtype='object')

In [4]:
bos.columns[50:100]

Index(['Climate_Priority', 'Climate_Risk', 'SMMA_FA_Visit_Date',
       'SMMA_FA_Status', 'SMMA_FA_School_Enrollment',
       'SMMA_FA_Enrollment_Building', 'SMMA_FA_Major_Investments',
       'SMMA_FA_Existing_PV', 'SMMA_FA_space_Solar', 'SMMA_FA_Roof_Membrane',
       'SMMA_FA_Facade', 'SMMA_FA_Windows', 'SMMA_FA_Boilers',
       'SMMA_FA_Heating_Distribution_Systems',
       'SMMA_FA_Ventilation_Distribution_Systems',
       'SMMA_FA_Electrical_Service', 'SMMA_FA_Means_Egress',
       'SMMA_FA_FP_Sprinklers', 'SMMA_FA_Life_Safety',
       'SMMA_FA_Security_Entry', 'SMMA_FA_Lighting_Quantity_Control',
       'SMMA_FA_Toilets_Fixtures', 'SMMA_FA_Plumbing_Distribution_Systems',
       'SMMA_FA_Accessibility', 'SMMA_FA_Deterioration_Roof',
       'SMMA_FA_Deterioration_Floor', 'SMMA_FA_Deterioration_Walls_Columns',
       'SMMA_FA_Deterioration_Foundations', 'SMMA_FA_Deterioration_Facade',
       'SMMA_FA_lateral_system_identifiable',
       'SMMA_FA_Overall_Building_Condition', 'SMMA_F

In [5]:
bos.columns[100:150]

Index(['SMMA_FA_Walls_Slopes', 'SMMA_FA_Overall_Site_Condition',
       'SMMA_FA_MHC_Archaeological_Assets', 'SMMA_FA_Mass_Transit',
       'SMMA_FA_Bikable', 'SMMA_FA_Walkable', 'SMMA_FA_Overall_Community_Site',
       'SMMA_FA_Site_suitability_school', 'Visit_Date_Date', 'SMMA_ES_Status',
       'SMMA_EA_Building_originally_designed',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_1',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_3',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_5',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_6',
       'SMMA_EA_grade_configuration_best_suited_4_to_6',
       'SMMA_EA_grade_configuration_best_suited_6_to_8',
       'SMMA_EA_grade_configuration_best_suited_7_to_8',
       'SMMA_EA_grade_configuration_best_suited_6_to_12',
       'SMMA_EA_grade_configuration_best_suited_7_to_12',
       'SMMA_EA_grade_configuration_best_suited_9_to_12',
       'SMMA_EA_Building_Ventilation', 'SMMA_EA_Building_Natural_Daylighting

In [6]:
bos.columns[150:200]

Index(['SMMA_EA_ES_Adequacy_Music', 'SMMA_EA_ES_Adequacy_Gymnasium',
       'SMMA_EA_ES_Adequacy_Media_Center', 'SMMA_EA_ES_Adequacy_Cafeteria',
       'SMMA_EA_ES_Adequacy_Stage', 'SMMA_EA_ES_Adequacy_Medical',
       'SMMA_EA_ES_Adequacy_Administration', 'SMMA_EA_ES_Adequacy_Custodial',
       'SMMA_EA_ES_Adequacy_Network', 'SMMA_EA_ES_Adequacy_Other_1',
       'SMMA_EA_ES_Adequacy_Other_2', 'SMMA_EA_ES_Adequacy_Other_3',
       'SMMA_EA_ES_Overall_EFE_spaces', 'SMMA_EA_MS_Adequacy_Classroom',
       'SMMA_EA_MS_Adequacy_Science', 'SMMA_EA_MS_Adequacy_SPED_SS',
       'SMMA_EA_MS_Adequacy_SPED_Resource', 'SMMA_EA_MS_Adequacy_Art',
       'SMMA_EA_MS_Adequacy_Music', 'SMMA_EA_MS_Adequacy_Vocations',
       'SMMA_EA_MS_Adequacy_Gymnasium', 'SMMA_EA_MS_Adequacy_Media_Center',
       'SMMA_EA_MS_Adequacy_Cafeteria', 'SMMA_EA_MS_Adequacy_Stage',
       'SMMA_EA_MS_Adequacy_Medical', 'SMMA_EA_MS_Adequacy_Administration',
       'SMMA_EA_MS_Adequacy_Custodial', 'SMMA_EA_MS_Adequacy_Network'

In [7]:
bos.columns[200:252]

Index(['SMMA_EA_HS712_Adequacy_Network', 'SMMA_EA_HS712_Adequacy_Other_1',
       'SMMA_EA_HS712_Adequacy_Other_2', 'SMMA_EA_HS712_Adequacy_Other_3',
       'SMMA_EA_HS712_Overall_EFE_spaces', 'SMMA_EA_HS912_Adequacy_Classroom',
       'SMMA_EA_HS912_Adequacy_TeacherPlanning',
       'SMMA_EA_HS912_Adequacy_Small_Group', 'SMMA_EA_HS912_Adequacy_Science',
       'SMMA_EA_HS912_Adequacy_SPED_SS',
       'SMMA_EA_HS912_Adequacy_SPED_Resource', 'SMMA_EA_HS912_Adequacy_Art',
       'SMMA_EA_HS912_Adequacy_Music', 'SMMA_EA_HS912_Adequacy_Vocations',
       'SMMA_EA_HS912_Adequacy_Gymnasium',
       'SMMA_EA_HS912_Adequacy_GynasiumAlternatives',
       'SMMA_EA_HS912_Adequacy_Media_Center',
       'SMMA_EA_HS912_Adequacy_Cafeteria', 'SMMA_EA_HS912_Adequacy_Auditorium',
       'SMMA_EA_HS912_Adequacy_Stage', 'SMMA_EA_HS912_Adequacy_Medical',
       'SMMA_EA_HS912_Adequacy_Administration',
       'SMMA_EA_HS912_Adequacy_Custodial', 'SMMA_EA_HS912_Adequacy_Network',
       'SMMA_EA_HS912_Adequac

#### Where do we begin?

One good place to be would be with the `NaaNs`. If there are any columns with a large about of data missing (i.e., 50% of more) -- dropping these columns from the dataset could be a good place to start.

In [8]:
# Let's look at 100 columns with the most NaaNs
bos.isna().sum().sort_values(ascending = False)[:100]

SMMA_EA_HS712_Adequacy_Network                    141
SMMA_EA_grade_configuration_best_suited_4_to_6    141
SMMA_EA_MS_Adequacy_Network                       141
SMMA_EA_HS712_Adequacy_Other_3                    141
SMMA_EA_K8_Adequacy_Network                       140
                                                 ... 
SMMA_EA_K8_Adequacy_Media_Center                  105
SMMA_EA_K8_Adequacy_Cafeteria                     105
SMMA_EA_K8_Adequacy_Stage                         105
SMMA_EA_K8_Adequacy_SPED_SS                       105
SMMA_EA_K8_Adequacy_Medical                       105
Length: 100, dtype: int64

100 columns in the dataset have **75% or more** missing data! The data in these columns may have been valuable, but I wouldn't feel comfortable filling `NaaNs` with any values (i.e., mean value of the column) with so much data missing. Let's drop these columns from the dataset.

In [9]:
[column for column in bos.columns if bos[column].isna().sum() >= 105]

['SMMA_Only_For_Map',
 'MSBA_Year_Reno',
 'SMMA_EA_grade_configuration_best_suited_4_to_6',
 'SMMA_EA_grade_configuration_best_suited_6_to_8',
 'SMMA_EA_grade_configuration_best_suited_7_to_8',
 'SMMA_EA_grade_configuration_best_suited_6_to_12',
 'SMMA_EA_grade_configuration_best_suited_7_to_12',
 'SMMA_EA_grade_configuration_best_suited_9_to_12',
 'SMMA_EA_ES_Adequacy_Custodial',
 'SMMA_EA_ES_Adequacy_Network',
 'SMMA_EA_ES_Adequacy_Other_1',
 'SMMA_EA_ES_Adequacy_Other_2',
 'SMMA_EA_ES_Adequacy_Other_3',
 'SMMA_EA_MS_Adequacy_Classroom',
 'SMMA_EA_MS_Adequacy_Science',
 'SMMA_EA_MS_Adequacy_SPED_SS',
 'SMMA_EA_MS_Adequacy_SPED_Resource',
 'SMMA_EA_MS_Adequacy_Art',
 'SMMA_EA_MS_Adequacy_Music',
 'SMMA_EA_MS_Adequacy_Vocations',
 'SMMA_EA_MS_Adequacy_Gymnasium',
 'SMMA_EA_MS_Adequacy_Media_Center',
 'SMMA_EA_MS_Adequacy_Cafeteria',
 'SMMA_EA_MS_Adequacy_Stage',
 'SMMA_EA_MS_Adequacy_Medical',
 'SMMA_EA_MS_Adequacy_Administration',
 'SMMA_EA_MS_Adequacy_Custodial',
 'SMMA_EA_MS_Adequac

In [10]:
# Drop columns with over 105 NaaNs from dataset
bos = bos.drop(columns = [column for column in bos.columns if bos[column].isna().sum() >= 105])

# Check work
bos.isna().sum().sort_values(ascending = False)

SMMA_EA_K8_Overall_EFE_spaces                        104
SMMA_EA_grade_configuration_best_suited_PreK_to_6    104
Climate_Risk                                          99
Climate_Priority                                      97
SMMA_EA_ES_Adequacy_Art                               92
                                                    ... 
SMMA_FA_Enrollment_Building                            0
SMMA_FA_School_Enrollment                              0
SMMA_FA_Status                                         0
 BPS_Total_Energy_Cost                                 0
SMMA_Identifier                                        0
Length: 151, dtype: int64

Even after dropping 100 columns from our dataset that were riddled with `NaaNs`, we still have 151 columns, some of which still have over 50% `NaaNs` in them (see above). Let's take a look at ant columns with over `50%` of their values missing from the dataset.

In [11]:
[column for column in bos.columns if bos[column].isna().sum() >= bos.shape[0] / 2]

['Climate_Priority',
 'Climate_Risk',
 'SMMA_EA_grade_configuration_best_suited_PreK_to_5',
 'SMMA_EA_grade_configuration_best_suited_PreK_to_6',
 'SMMA_EA_ES_Adequacy_PreK',
 'SMMA_EA_ES_Adequacy_Kindergarten',
 'SMMA_EA_ES_Adequacy_Classroom',
 'SMMA_EA_ES_Adequacy_SPED_SS',
 'SMMA_EA_ES_Adequacy_SPED_Resource',
 'SMMA_EA_ES_Adequacy_Art',
 'SMMA_EA_ES_Adequacy_Music',
 'SMMA_EA_ES_Adequacy_Gymnasium',
 'SMMA_EA_ES_Adequacy_Media_Center',
 'SMMA_EA_ES_Adequacy_Cafeteria',
 'SMMA_EA_ES_Adequacy_Stage',
 'SMMA_EA_ES_Adequacy_Medical',
 'SMMA_EA_ES_Adequacy_Administration',
 'SMMA_EA_ES_Overall_EFE_spaces',
 'SMMA_EA_K8_Overall_EFE_spaces']

In [12]:
# Drop columns with over 50% NaaNs from dataset
bos = bos.drop(columns = [column for column in bos.columns if bos[column].isna().sum() >= bos.shape[0] / 2])

# Check work
bos.isna().sum().sort_values(ascending = False)

SMMA_EA_grade_configuration_best_suited_PreK_to_3    64
SMMA_EA_grade_configuration_best_suited_PreK_to_1    56
SMMA_EA_Building_originally_designed                 18
DOE_SP                                               17
BPS_Year_Founded                                     17
                                                     ..
SMMA_FA_Overall_Building_Condition                    0
SMMA_FA_Overall_Site_Condition                        0
SMMA_FA_Overall_Community_Site                        0
SMMA_EA_Overall_EFE_learning_environments             0
SMMA_Identifier                                       0
Length: 132, dtype: int64

Ok...last command in which `NaaNs`, I promise.

After looking at the columns that remain in our dataset, we can see that there are still two columns with `64` and `56` `NaaN` values. After that, the next highest `NaaN` count is `18`, which is much more managable.

Let's drop those other 2 columns from our dataset.

In [13]:
# Drop remaining two columns with over 50 NaaNs
bos.drop(columns = ['SMMA_EA_grade_configuration_best_suited_PreK_to_3',
                    'SMMA_EA_grade_configuration_best_suited_PreK_to_1'],
         inplace = True)

In [14]:
# How many columns do we have now?
bos.shape

(141, 130)

There are still 130 columns in this dataset, which is a lot to include for just 141 schools. If we drop any more columns from this dataset, we'll have to be more objective.

In [15]:
bos[bos.columns[:20]]

SMMA_Identifier                       BPS_School_Name  \
0               031              Adams, Samuel Elementary   
1               078    Alighieri, Dante Montessori School   
2              045A            Another Course to College*   
3               012  Baldwin Early Learning Pilot Academy   
4               087             Bates, Phineas Elementary   
..              ...                                   ...   
136            011B                 West Roxbury Academy*   
137            047B      West Zone Early Learning Center*   
138             100          Winship, F. Lyman Elementary   
139             073             Winthrop, John Elementary   
140             044    Young Achievers Science & Math K-8   

            BPS_Historical_Name SMMA_Abbreviated_Name  \
0                         Adams                 Adams   
1                     Alighieri             Alighieri   
2                          Taft                  ACC*   
3                   Baldwin ELC               Baldwin   
4                         Bates                 Bates   
..                          ...                   ...   
136    West Roxbury High School         West Roxbury*   
137  Hennigan Elementary School            West Zone*   
138                     Winship               Winship   
139         Winthrop Elementary              Winthrop   
140                   Lewenberg       Young Achievers   

                               BPS_Address BRA_Neighborhood  SMMA_latitude  \
0     165 Webster St East Boston, MA 02128      East Boston      42.365553   
1     37 Gove Street East Boston, MA 02128      East Boston      42.371565   
2      20 Warren Street Brighton, MA 02135          Allston      42.350354   
3          121 Corey Rd Brighton, MA 02135         Brighton      42.342037   
4        426 Beech St Roslindale, MA 02131       Roslindale      42.277663   
..                                     ...              ...            ...   
136  1205 VFW Pkwy, West Roxbury, MA 02132     West Roxbury      42.282206   
137   200 Heath St Jamaica Plain, MA 02130    Jamaica Plain      42.326026   
138       54 Dighton St Brighton, MA 02135         Brighton      42.347723   
139   35 Brookford St Dorchester, MA 02125          Roxbury      42.318387   
140        20 Outlook Rd Mattapan, MA02126         Mattapan      42.281976   

     SMMA_longitude      SMMA_Typology SMMA_Educational_Program  \
0        -71.034917  Elementary School   Elementary: Pre-K to 6   
1        -71.037608  Elementary School   Elementary: Pre-K to 6   
2        -71.145582        High School    High Schools: 9 to 12   
3        -71.140529     Early Learning   Elementary: Pre-K to 6   
4        -71.135353  Elementary School   Elementary: Pre-K to 6   
..              ...                ...                      ...   
136      -71.174549        High School    High Schools: 9 to 12   
137      -71.106821     Early Learning   Elementary: Pre-K to 6   
138      -71.154965  Elementary School   Elementary: Pre-K to 6   
139      -71.075341  Elementary School   Elementary: Pre-K to 6   
140      -71.093909                K-8   Elementary: Pre-K to 8   

    BPS_Grades_Offered BPS_Property_Status  BPS_Total_GSF   BPS_GSF  \
0                 K0-5               Owned        41336.0   41336.0   
1                 K0-5               Owned        16948.0   16948.0   
2                 9-12               Owned        28021.0   28021.0   
3                 K0-1               Owned        27879.0   27879.0   
4                 K1-5               Owned        27936.0   27936.0   
..                 ...                 ...            ...       ...   
136               9-12               Owned       113508.0  113508.0   
137               K0-1               Owned        27000.0   27000.0   
138               K0-5               Owned        41016.0   41016.0   
139               K1-5               Owned        40530.0   40530.0   
140               K1-8               Owned       120944.0  1209

**Let's drop the following columns:**

- `SMMA_Identifier` - we're going to use `BPS_School_Name` as the index, so there's no reason to include this column. 
- `BPS_Historical_Name` - not necessary. This noise in addition to the `BPS_School_Name` column.
- `SMMA_Abbreviated_Name` - also not necessary for the same reason above
- `BPS_Address` - not important because the longitudinal and latidudinal data is included in the dataset
- `BPS_Total_SF` - not necessary, a redundant value of the `BPS_GSF` column
- `SMMA_Site_SF` - redundant of the `SMMA_Site_Acres` column
- `SMMA_Era` - not necessary

In [16]:
bos.drop(columns = ['SMMA_Identifier', 'BPS_Historical_Name', 'SMMA_Abbreviated_Name', 
                    'BPS_Address', 'BPS_Total_GSF', 'SMMA_Site_SF', 'SMMA_Era'], inplace = True)

bos[bos.columns[:20]]

BPS_School_Name BRA_Neighborhood  SMMA_latitude  \
0                Adams, Samuel Elementary      East Boston      42.365553   
1      Alighieri, Dante Montessori School      East Boston      42.371565   
2              Another Course to College*          Allston      42.350354   
3    Baldwin Early Learning Pilot Academy         Brighton      42.342037   
4               Bates, Phineas Elementary       Roslindale      42.277663   
..                                    ...              ...            ...   
136                 West Roxbury Academy*     West Roxbury      42.282206   
137      West Zone Early Learning Center*    Jamaica Plain      42.326026   
138          Winship, F. Lyman Elementary         Brighton      42.347723   
139             Winthrop, John Elementary          Roxbury      42.318387   
140    Young Achievers Science & Math K-8         Mattapan      42.281976   

     SMMA_longitude      SMMA_Typology SMMA_Educational_Program  \
0        -71.034917  Elementary School   Elementary: Pre-K to 6   
1        -71.037608  Elementary School   Elementary: Pre-K to 6   
2        -71.145582        High School    High Schools: 9 to 12   
3        -71.140529     Early Learning   Elementary: Pre-K to 6   
4        -71.135353  Elementary School   Elementary: Pre-K to 6   
..              ...                ...                      ...   
136      -71.174549        High School    High Schools: 9 to 12   
137      -71.106821     Early Learning   Elementary: Pre-K to 6   
138      -71.154965  Elementary School   Elementary: Pre-K to 6   
139      -71.075341  Elementary School   Elementary: Pre-K to 6   
140      -71.093909                K-8   Elementary: Pre-K to 8   

    BPS_Grades_Offered BPS_Property_Status   BPS_GSF  SMMA_Site_Acres  \
0                 K0-5               Owned   41336.0         0.933884   
1                 K0-5               Owned   16948.0         0.505946   
2                 9-12               Owned   28021.0         1.907438   
3                 K0-1               Owned   27879.0         0.688476   
4                 K1-5               Owned   27936.0         1.539876   
..                 ...                 ...       ...              ...   
136               9-12               Owned  113508.0        44.499495   
137               K0-1               Owned   27000.0         5.672222   
138               K0-5               Owned   41016.0         0.788246   
139               K1-5               Owned   40530.0         0.846166   
140               K1-8               Owned  120944.0         4.017378   

    BPS_Year_Founded   BPS_Year_Built              BPS_Plans  \
0                1910          1910.0  CAD plans, site plans   
1                1924          1924.0  CAD plans, site plans   
2                1939          1895.0  CAD plans, site plans   
3                1926          1926.0  CAD plans, site plans   
4                1929          1929.0  CAD plans, site plans   
..                ...             ...                    ...   
136              2011          1976.0  CAD plans, site plans   
137              1997          1972.0  CAD plans, site plans   
138              1899          1901.0  CAD plans, site plans   
139              1911          1911.0  CAD plans, site plans   
140              1995          1930.0  CAD plans, site plans   

          BPS_Principal    BPS_Open   BPS_Close    DOE_ID  DOE_Total  DOE_PK  \
0          Hannah Irvin  7:30:00 AM  1:30:00 PM  350302.0        302    43.0   
1          Mayra Cuevas  7:30:00 AM  1:30:00 PM  350066.0         90    29.0   
2    Lisa Gilbert-Smith  7:30:00 AM  2:30:00 PM  350541.0        237     0.0   
3            Tavia Mead  7:30:00 AM  4:35:00 PM  350003.0        142    68.0   
4            Kelly Hung  9:30:00 AM  3:30:00 PM  350278.0        297    20.0   
..                  ...         ...         ...       ...        ...     ...   
136                 NaN         NaN         NaN  350574.0        529     0.0   
137       

In [17]:
bos[bos.columns[16:36]]

DOE_ID  DOE_Total  DOE_PK  DOE_K  DOE_1  DOE_2  DOE_3  DOE_4  DOE_5  \
0    350302.0        302    43.0   49.0   53.0   42.0   48.0   37.0   30.0   
1    350066.0         90    29.0   15.0   16.0   12.0    8.0    6.0    4.0   
2    350541.0        237     0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3    350003.0        142    68.0   41.0   33.0    0.0    0.0    0.0    0.0   
4    350278.0        297    20.0   36.0   39.0   44.0   49.0   58.0   51.0   
..        ...        ...     ...    ...    ...    ...    ...    ...    ...   
136  350574.0        529     0.0    0.0    0.0    0.0    0.0    0.0    0.0   
137  350006.0         93    24.0   34.0   35.0    0.0    0.0    0.0    0.0   
138  350374.0        258    23.0   33.0   42.0   44.0   49.0   34.0   33.0   
139  350180.0        349    28.0   60.0   50.0   59.0   65.0   48.0   39.0   
140  350380.0        542    34.0   48.0   51.0  115.0   64.0   63.0   34.0   

     DOE_6  DOE_7  DOE_8  DOE_9  DOE_10  DOE_11  DOE_12  DOE_SP  \
0      0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0   
1      0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0   
2      0.0    0.0    0.0   54.0    62.0    62.0    59.0     0.0   
3      0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0   
4      0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0   
..     ...    ...    ...    ...     ...     ...     ...     ...   
136    0.0    0.0    0.0  136.0   140.0   103.0   147.0     3.0   
137    0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0   
138    0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0   
139    0.0    0.0    0.0    0.0     0.0     0.0     0.0     0.0   
140   42.0   47.0   44.0    0.0     0.0     0.0     0.0     0.0   

     GSF / Student Value    BPS_KWH  BPS_Electric_Bill   
0                  180.0   381000.0         $58,339.00   
1                  180.0     9977.0          $2,072.00   
2                  225.0   343410.0         $49,657.00   
3                  200.0    70212.0          $9,588.00   
4                  180.0    91800.0         $13,367.00   
..                   ...        ...                 ...  
136                225.0  1097520.0        $294,332.00   
137                200.0   120384.0         $74,435.00   
138                180.0   128480.0         $20,522.00   
139                175.0   138240.0         $20,594.00   
140                195.0   385248.0         $53,897.00   

[141 rows x 20 columns]

In [18]:
bos[bos.columns[35:56]]

BPS_Electric_Bill   BPS_Therms  BPS_Gas_Bill   BPS_Water_ft_cubic  \
0           $58,339.00      14650.0    $13,401.00                  0.0   
1            $2,072.00       8230.0     $9,957.00               8120.0   
2           $49,657.00      51459.0    $57,978.00             163110.0   
3            $9,588.00       9806.0    $11,657.00              42550.0   
4           $13,367.00      14530.0    $16,909.00              42870.0   
..                  ...         ...            ...                 ...   
136        $294,332.00     257019.0   $258,704.00             265450.0   
137         $74,435.00     267450.0   $276,004.00             287880.0   
138         $20,522.00      15656.0    $18,100.00              44180.0   
139         $20,594.00      16897.0    $19,484.00              53440.0   
140         $53,897.00      36421.0    $56,973.00              90940.0   

     BPS_Water_Bill   BPS_Total_Energy_Cost  SMMA_FA_Visit_Date  \
0              $-                $71,740.00          05/18/2016   
1         $1,685.00              $13,714.00          05/18/2016   
2        $20,763.00             $128,398.00          04/27/2016   
3         $5,227.00              $26,472.00          04/25/2016   
4         $5,262.00              $35,538.00          06/08/2016   
..               ...                     ...                ...   
136      $36,183.00             $589,219.00          06/01/2016   
137      $37,653.00             $388,092.00          04/26/2016   
138       $5,394.00              $44,016.00          05/10/2016   
139       $7,457.00              $47,535.00          06/10/2016   
140      $11,457.00             $122,327.00          06/11/2016   

    SMMA_FA_Status  SMMA_FA_School_Enrollment  SMMA_FA_Enrollment_Building  \
0          Phase 3                        302                          302   
1          Phase 3                         90                           90   
2          Phase 3                        237                          674   
3          Phase 3                        142                          142   
4          Phase 3                        297                          297   
..             ...                        ...                          ...   
136        Phase 3                        529                         1026   
137        Phase 3                         93                          717   
138        Phase 3                        258                          258   
139        Phase 3                        349                          349   
140        Phase 3                        542                          542   

     ... SMMA_FA_Existing_PV SMMA_FA_space_Solar SMMA_FA_Roof_Membrane  \
0    ...                  No                 Yes              Moderate   
1    ...                  No                 Yes                 Minor   
2    ...                  No                 Yes                 Minor   
3    ...                  No                 Yes           Not Present   
4    ...                  No                 Yes              Moderate   
..   ...                 ...                 ...                   ...   
136  ...                  No                 Yes               Replace   
137  ...                  No                 Yes              Moderate   
138  ...                  No                 Yes              Moderate   
139  ...                  No                 Yes                 Minor   
140  ...                  No                 Yes                 Minor   

    SMMA_FA_Facade SMMA_FA_Windows SMMA_FA_Boilers  \
0         Adequate         Replace         Replace   
1         Adequate        Adequate        Adequate   
2         Adequate           Minor        Moderate   
3            Minor           Minor        Moderate   
4         Adequate        Moderate         Replace   
..             ...             ...             ...   
136       Adequate         Replace        Adequate   
137       Adequate           Minor        Adequate   

In [19]:
bos[' BPS_Electric_Bill ']

0        $58,339.00 
1         $2,072.00 
2        $49,657.00 
3         $9,588.00 
4        $13,367.00 
           ...      
136     $294,332.00 
137      $74,435.00 
138      $20,522.00 
139      $20,594.00 
140      $53,897.00 
Name:  BPS_Electric_Bill , Length: 141, dtype: object

We've got a problem with the above column. For one, the column name has two spaces in it which made it tricky to isolate (some guess and check was involved). Also, we have a valuable quantitative column which has been saved in an `object` format, and with a `$` symbol that further complicates these values. We'll have to drop the `$` from each cell and change the data type for each cell from `object` to `float`.

In [20]:
# Rename column -- the name of this column has unecessary spaces in it
bos.rename(columns = {' BPS_Electric_Bill ' : 'BPS_Electric_Bill'}, inplace = True)

In [181]:
# Define function to change values from objects to floats
def object_currency_to_float(column):
    
    # Create empty list
    new_column = []
    
    # Loop through each cell in the inputted column
    for cell in bos[column]:
        
        # Check to see if each cell is a string
        if type(cell) == str:
            
            cell = cell.replace('$', '')
    
            cell = cell.replace('-', '0')
            
            cell = cell.replace(',', '')
                
            cell = float(cell)
            
            new_column.append(cell)
            
        else:
            
            new_column.append(cell)
    
    return new_column
                
                    
    

In [172]:
# Call function and set equal to the Electric Bill column
bos['BPS_Electric_Bill'] = object_currency_to_float('BPS_Electric_Bill')

In [51]:
# Fill NaaNs in this column with the average value of Electric Bills that we have available 
#\n in the dataset
bos['BPS_Electric_Bill'].fillna(bos['BPS_Electric_Bill'].mean(), inplace = True)

It's become clear that there are unecessary spaces in multiple `column` names for this dataset. Instead of changing them all one by one, let's run the command below to get rid of them all in one fell swoop.

In [72]:
for column in bos.columns:
    bos.rename(columns = {column : column.replace(' ', '')}, inplace = True)

Let's check to see if we'll have to run the `object_currency_to_float` function for other columns (Hint: we will.).

In [159]:
bos['BPS_Gas_Bill']

0        $13,401.00 
1         $9,957.00 
2        $57,978.00 
3        $11,657.00 
4        $16,909.00 
           ...      
136     $258,704.00 
137     $276,004.00 
138      $18,100.00 
139      $19,484.00 
140      $56,973.00 
Name: BPS_Gas_Bill, Length: 141, dtype: object

In [182]:
bod[]'Bosobject_currency_to_float('BPS_Gas_Bill')

[13401.0,
 9957.0,
 57978.0,
 11657.0,
 16909.0,
 20307.0,
 124712.0,
 5269.0,
 66101.0,
 64059.0,
 27214.0,
 57978.0,
 36426.0,
 112539.0,
 181411.0,
 31297.0,
 13768.0,
 70999.0,
 84952.0,
 0.0,
 10694.0,
 167987.0,
 12811.0,
 10654.0,
 12706.0,
 59149.0,
 148439.0,
 16293.0,
 16031.0,
 41436.0,
 nan,
 109138.0,
 nan,
 61154.0,
 11588.0,
 15931.0,
 122202.0,
 60369.0,
 39855.0,
 20511.0,
 0.0,
 nan,
 30013.0,
 18191.0,
 nan,
 197335.0,
 16350.0,
 109794.0,
 38544.0,
 52294.0,
 25072.0,
 nan,
 8287.0,
 31525.0,
 8779.0,
 16084.0,
 11742.0,
 18247.0,
 13910.0,
 73556.0,
 21587.0,
 32237.0,
 60996.0,
 276004.0,
 39700.0,
 13718.0,
 36674.0,
 22583.0,
 258136.0,
 25977.0,
 59636.0,
 258136.0,
 nan,
 14013.0,
 32656.0,
 10565.0,
 12224.0,
 9830.0,
 nan,
 55473.0,
 17155.0,
 131602.0,
 148757.0,
 26268.0,
 7083.0,
 nan,
 11948.0,
 15819.0,
 12731.0,
 108616.0,
 109138.0,
 23804.0,
 7185.0,
 41112.0,
 11612.0,
 16530.0,
 76238.0,
 95706.0,
 9515.0,
 95706.0,
 118450.0,
 64059.0,
 nan,
 8488